### DISCUSSION TITLE
Tuya MCU communication problem

### DISCUSSION THREAD

**Discussion by gooman-uk** [arendst/Tasmota#10470] [Q&A] [answered]
I have a panel heater that utilises a Tuya board for wifi control. I've successfully flashed it with Tasmota 9.2, but can't communicate with the MCU.

Here's what I'm seeing in the console after enabling logging:
```
11:32:51 APP: Boot Count 12
11:32:51 CFG: Saved to flash at F7, Count 29, Bytes 4096
11:32:52 MQT: tele/tasmota_B7AE5F/STATE = {"Time":"2021-01-09T11:32:52","Uptime":"0T00:00:11","UptimeSec":11,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"OFF","Wifi":{"AP":1,"SSId":"PLUSNET-WX58","BSSId":"D0:57:94:C4:4B:72","Channel":3,"RSSI":52,"Signal":-74,"LinkCount":1,"Downtime":"0T00:00:05"}}
11:32:52 TYA: Send "55aa00000000ff"
11:32:57 MQT: tasmota/discovery/ECFABCB7AE5F/config = {"ip":"192.168.1.113","dn":"Tasmota-Off-Heater","fn":["Tasmota-Off-Heater",null,null,null,null,null,null,null],"hn":"tasmota_B7AE5F-3679","mac":"ECFABCB7AE5F","md":"Tuya MCU","ty":1,"if":0,"ofln":"Offline","onln":"Online","state":["OFF","ON","TOGGLE","HOLD"],"sw":"9.2.0","t":"tasmota_B7AE5F","ft":"%prefix%/%topic%/","tp":["cmnd","stat","tele"],"rl":[1,0,0,0,0,0,0,0],"swc":[-1,-1,-1,-1,-1,-1,-1,-1],"swn":[null,null,null,null,null,null,null,null],"btn":[0,0,0,0],"so":{"4":0,"11":0,"13":0,"17":0,"20":0,"30":0,"68":0,"73":0,"82":0,"114":0},"lk":1,"lt_st":0,"sho":[0,0,0,0],"ver":1} (retained)
11:32:57 MQT: tasmota/discovery/ECFABCB7AE5F/sensors = {"sn":{"Time":"2021-01-09T11:32:57"},"ver":1} (retained)
11:33:03 TYA: Send "55aa00000000ff"
11:33:06 WIF: Checking connection...
11:33:14 TYA: Send "55aa00000000ff"
11:33:25 TYA: Send "55aa00000000ff"
11:33:26 WIF: Checking connection...
11:33:36 TYA: Send "55aa00000000ff"
11:33:46 WIF: Checking connection...
11:33:47 TYA: Send "55aa00000000ff"
11:33:58 TYA: Send "55aa00000000ff"
```

The Tuya chip is a TYWE1S on a TYJW2S-5V board (see https://developer.tuya.com/en/docs/iot/device-development/module/wifi-module/tyjw-series-module/wifijw2s5vmodule?id=K9605srhjahvz)

I've tried the Tuya Tx/Rx on both GPIO3/1 and 15/13, but with no success.

Any suggestions?

**Comment by sfromis** [comment, marked as answer]
Try https://tasmota.github.io/docs/Commands/#setoption97

**Reply by gooman-uk** [reply]
Thanks for the pointer. Just tried with setoption97 as 0 and 1 on both GPIO configurations ... still no joy :-(

**Comment by gooman-uk** [comment]
OK, in the absence of any other suggestions I'm thinking of the next steps as:

1. Solder wires onto the MCU comms port on the TYJW2S-5V board and connect it to my FTDI USB stick providing 5V power, and see if I'm getting anything out
2. If I am, use the Tuya Module Debugging tool (which can simulate an MCU) to see if it's what the MCU expects
3. If that doesn't shed any light, repeat the same process with the MCU (as the Tuya Module Debugging tool can also simulate the TYJW2S-5V board)

If step 1 fails then we have an issue where neither GPIOs 3/1 or 15/13 are used as Tx/Rx at either 9600 or 15200 baud. Which would be a real head-scratcher!

**Comment by jodibb** [comment]
Hi. I have similar issue. Also using TYWE1S and 9.2.0 on diffuser with same result. I already tried same as you, set different gpio and baud rate. But with same result. No communication. I was checking with multimeter and it looks like GPIO1/3 should be correct. 
Did you find any solution?

> 00:52:31 TYA: Send "55aa00000000ff"
00:52:42 TYA: Send "55aa00000000ff"
00:52:48 WIF: Checking connection...
00:52:53 TYA: Send "55aa00000000ff"
00:53:04 TYA: Send "55aa00000000ff"
00:53:08 WIF: Checking connection...
00:53:15 TYA: Send "55aa00000000ff"
00:53:26 TYA: Send "55aa00000000ff"

**Reply by gooman-uk** [reply]
My device was Tx on 13 and Rx on 15. To cap that, my devices uses a completely non-standard protocol between the MCU and ESP, so I couldn't use the included Tuya functionality in Tasmota. Rather than DpIDs and discrete parameters it embeds all the parameters into a single hex string.

I ended up creating a new Template, writing a bunch of Rules, and an MQTT helper app to translate between the hex string and discrete MQTT parameters.

See https://templates.blakadder.com/devola_designer.html for details.

**Reply by jodibb** [reply]
Thanks. I'll check that. I thing I'll have similar problem as you and it will use non-standard protocol.

**Reply by gooman-uk** [reply]
Basically, once I'd identified the correct GPIOs for the serial data, we had to make a serial sniffer to monitor the traffic between the MCU and stock firmware.

**Reply by jodibb** [reply]
Sadly I'm not that far, to be able make sniffer. I think I have correct GPIOs. I was checking it with mutlimeter and it's connected to MCU. But no response. According to templates, most of diffusers have almost same serial codes, so I'll try them.

**Reply by jodibb** [reply]
Ok, solved. I did the same as you. Serial sniffer to monitor traffic. But sadly, I had tasmota flashed, so I was able to find only reports back to tuya. Anyway, once I wanted to flash back original firmware, I pass again though pinout again, and I found out the GPIO was wrong and this module is using 13/15, as yours. And now, everything works. Weird is, that I was trying 13/15 before. But without success. I don't understand, why now, after all soldering buying serial analyzer and spending X hours on debuging, it started to work. :)

**Reply by Greefon** [reply]
Hi
I have the same TYWE1S with non standard protocol for MCU which I read and checked by usb-ttl (I used original firmware to get all working commands).
But I can't make TYWE1S to send anything from U1TX pin.
How did you set up the tasmota to make it serialsend ?
![2021-02-07_17-51-50](https://user-images.githubusercontent.com/16508296/107152337-09680e00-6970-11eb-82fc-c8bcd6521735.png)

**Reply by Greefon** [reply]
I made it.
GPIO13 - SerBr Rx
GPIO15 - SerBr Tx

close
SSerialSend5%2DFD3270255FEFE0302F925 
stop
SSerialSend5%2DFD355FEFE030338E5
open
SSerialSend5%2DFD3012755FEFE0301B924
curtain motor holfmann (dooya) MT82TV
